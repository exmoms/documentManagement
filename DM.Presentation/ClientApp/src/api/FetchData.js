/**
 * @file this file includes functions that fetch data from API
 */

// Generic Delete Method
export async function deleteFromApi(url) {
  let res = await fetch(url, {
    method: "DELETE",
    headers: {
      Accept: "application/json",
      "Content-Type": "application/json",
    },
  });
  return res;
}

// Get MetaData Models
export async function fetchData(url) {
  let object = {};
  /*
   * send the request through a CORS proxy
   * https://stackoverflow.com/questions/43262121/trying-to-use-fetch-and-pass-in-mode-no-cors/43268098#43268098
   */
  await fetch(url, {
    method: "GET",
    headers: {
      Accept: "application/json",
      "Content-Type": "application/json",
    },
  })
    .then((response) => response.json())
    .then(
      (data) => (object = data),
      (error) => console.log(error)
    );
  return object;
}
