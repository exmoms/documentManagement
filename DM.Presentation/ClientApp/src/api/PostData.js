/**
 * @file This file includes functions that help to post data to a specific API
 * @version 1.0
 */

/**
 * Post new document data.
 * Upload form-data, scanned copy of doc and attachments.
 * @author Youssef Shaaban <y.shaaban@lit-co.net>
 * @param {*} documentData
 * @param {*} images
 * @param {*} attachments
 * @param {*} url
 * @param {*} option its values either "Add" for add new document version or "Update" to update exsiting document Version
 */
export async function postDocumentData(
  documentData,
  scannedPages,
  attachments,
  url = "",
  option
) {
  var formData = new FormData();

  if (process.env.NODE_ENV === "development") {
    console.log(documentData);
    console.log(attachments);
    console.log(scannedPages);
  }

  switch (option) {
    case "Add":
      formData.append("Document", JSON.stringify(documentData));
      break;
    case "Update":
      formData.append("Version", JSON.stringify(documentData));
      break;
    default:
  }

  attachments.forEach((attachment) => {
    formData.append("Attachments", attachment);
  });
  scannedPages.forEach((scannedPage) => {
    formData.append("Scans", scannedPage);
  });

  let response = {};
  await fetch(url, {
    // content-type header should not be specified!
    method: "POST",
    headers: {
      Accept: "application/json",
      // "Content-Type": "application/json",
    },
    body: formData,
  }).then(
    (res) => {
      response = res;
      // response.json();
    },
    (error) => console.log("[ERROR]:", error)
  );
  return response;
}

/**
 * Post Attachment Data.
 * Upload attachment json data + attachment file  .
 * @author Ali Daghman <ali.daghman@lit-co.net>
 * @param {*} payload
 * @param {*} attachmentFile
 * @param {*} url
 */
export async function postAttachmentData(payload, attachmentFile, url = "") {
  var formData = new FormData();

  formData.append("Attachment", JSON.stringify(payload));
  formData.append("FileAttachment", attachmentFile);

  let response = {};
  await fetch(url, {
    // content-type header should not be specified!
    method: "POST",
    body: formData,
  }).then(
    (res) => {
      response = res;
      // response.json();
    },
    (error) => console.log("[ERROR]:", error)
  );
  return response;
}

/**
 * Post Request To API
 * Send post request to api with  payload  and return a response.
 * @author Ali Daghman <ali.daghman@lit-co.net>
 * @param {*} url The api url
 * @param {*} payload js object which will be sent as request body.
 */
export async function postDataToAPI(url, payload = null, method = "POST") {

  const response = await fetch(url, {
    method: method,
    headers: {
      Accept: "application/json",
      "Content-Type": "application/json",
    },
    body: JSON.stringify(payload),
  });
  return response;
}
