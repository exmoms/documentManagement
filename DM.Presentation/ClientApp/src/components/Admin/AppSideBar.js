import Divider from "@material-ui/core/Divider";
import Drawer from "@material-ui/core/Drawer";
import Hidden from "@material-ui/core/Hidden";
import List from "@material-ui/core/List";
import ListItem from "@material-ui/core/ListItem";
import ListItemIcon from "@material-ui/core/ListItemIcon";
import ListItemText from "@material-ui/core/ListItemText";
import { makeStyles, useTheme } from "@material-ui/core/styles";
import AssignmentIcon from "@material-ui/icons/Assignment";
import DescriptionIcon from "@material-ui/icons/Description";
import FolderOpenIcon from "@material-ui/icons/FolderOpen";
import NoteAddIcon from "@material-ui/icons/NoteAdd";
import SearchIcon from "@material-ui/icons/Search";
import React from "react";
import { useTranslation } from "react-i18next";
import { Link } from "react-router-dom";
// Start: Style properties --------------
const drawerWidth = 260;
const useStyles = makeStyles((theme) => ({
  drawer: {
    [theme.breakpoints.up("sm")]: {
      width: drawerWidth,
      flexShrink: 0,
    },
  },
  toolbar: theme.mixins.toolbar,
  drawerPaper: {
    width: drawerWidth,
    // top: 65,
  },
  content: {
    flexGrow: 1,
    padding: theme.spacing(3),
  },
}));
//End --------------------------------------

// Start: function SideBar ---------------------------------------------------------------
function SideBar(props) {
  const { t } = useTranslation();
  const { container } = props;
  const classes = useStyles();
  const theme = useTheme();

  // list in sidebar ------------
  // Create Forms
  let createFormlist = [
    {
      path: "document-class",
      title: t("label_documentclass"),
      icon: <AssignmentIcon />,
    },
    {
      path: "metadata-model",
      title: t("createdocument_meatdatmodel"),
      icon: <DescriptionIcon />,
    },
  ];
  // -------------------------------------------

  // list in sidebar ------------
  // Full out forms
  let fullOutFormsList = [
    { path: "create-document", title: t("create_doc"), icon: <NoteAddIcon /> },
    {
      path: "browse-documents",
      title: t("browsedocuments"),
      icon: <SearchIcon />,
    },
    {
      path: "document-set",
      title: t("appsidebar_documentset"),
      icon: <FolderOpenIcon />,
    },
  ];
  // -------------------------------------------

  // Start: show the content of sideBar ---------
  const drawer = (
    <>
      <div>
        <div className={classes.toolbar} />
        {(props.user.role === "Admin" || props.user.role === "SuperAdmin") && (
          <React.Fragment>
            <List>
              {createFormlist.map((text, index) => (
                <ListItem
                  button
                  key={text.title}
                  component={Link}
                  to={"/" + text.path}
                >
                  <ListItemIcon>{text.icon}</ListItemIcon>
                  <ListItemText primary={text.title} />
                </ListItem>
              ))}
            </List>
            <Divider />
          </React.Fragment>
        )}
        <List>
          {fullOutFormsList.map((text, index) => (
            <ListItem
              button
              key={text.title}
              component={Link}
              to={"/" + text.path}
            >
              <ListItemIcon>{text.icon}</ListItemIcon>
              <ListItemText primary={text.title} />
            </ListItem>
          ))}
        </List>
      </div>
    </>
  );
  // End: -----------------------------------------------

  return (
    <nav
      className={classes.drawer}
      aria-label="mailbox folders"
      data-testid="app-side-bar"
    >
      {/* The implementation can be swapped with js to avoid SEO duplication of links. */}
      <Hidden smUp implementation="css">
        <Drawer
          id="drawer1"
          container={container}
          variant="temporary"
          anchor={theme.direction === "rtl" ? "right" : "left"}
          open={props.open}
          onClose={props.onClose}
          classes={{
            paper: classes.drawerPaper,
          }}
          ModalProps={{
            keepMounted: true, // Better open performance on mobile.
          }}
        >
          {drawer}
        </Drawer>
      </Hidden>{" "}
      <Hidden xsDown implementation="css">
        <Drawer
          id="drawer2"
          classes={{
            paper: classes.drawerPaper,
          }}
          variant="permanent"
          open
        >
          {drawer}
        </Drawer>
      </Hidden>
    </nav>
  );
}
// End: function SideBar ---------------------------------------------------------------

export default SideBar;
