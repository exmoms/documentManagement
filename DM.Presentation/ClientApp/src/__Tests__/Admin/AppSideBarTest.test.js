
import React from "react";
import ReactDom from "react-dom";
import AppSideBar from "../../components/Admin/AppSideBar";
import { cleanup } from "@testing-library/react";
import { BrowserRouter as Router } from "react-router-dom";
import { shallow , mount} from "enzyme";
import toJSON from "enzyme-to-json";
import { MemoryRouter } from "react-router-dom";
import renderer from "react-test-renderer";
import { Link } from "react-router-dom";
import { Suspense } from "react";
import * as FetchMock from "../../api/FetchData";
import * as PostDataMock from "../../api/PostData";
import * as ContextAuth from "../../context/auth";
import MenuItem from "@material-ui/core/MenuItem";
import { act } from "react-dom/test-utils";
import List from "@material-ui/core/List";
afterEach(cleanup);
describe("AppSideBar", () => {
   afterEach(() => {
    jest.clearAllMocks();
  });
  test("renders without crashing", () => {
    const div = document.createElement("div");
    const onClose = jest.fn();
    let user={role:"Admin"};
    let open = true;
    ReactDom.render(
        <Router> <AppSideBar user={user} open={open} onClose={onClose}/></Router>,
      div
    );
  });

  test("matches snapshot", () => {
    const onClose = jest.fn();
    let user={role:"Admin"};
    let open = true;
    const wrapper = shallow(
        <Router> <AppSideBar user={user} open={open} onClose={onClose}/></Router>
    );
    expect(toJSON(wrapper)).toMatchSnapshot();
  });
  
  test("intialization AppSideBar when role ia admin", () => {
    const onClose = jest.fn();
    let user={role:"Admin"};
    let open = true;
    let  wrapper = mount(<Router> <AppSideBar user={user} open={open} onClose={onClose}/></Router>) ;
    expect(wrapper.find("#drawer1").at(0).props().container).toBe();
    expect(wrapper.find("#drawer1").at(0).props().open).toBeTruthy();
    expect(wrapper.find(List).length).toBe(4);
  });

  test("intialization AppSideBar when role ia user", () => {
    const onClose = jest.fn();
    let user={role:"User"};
    let open = true;
    let  wrapper = mount(<Router> <AppSideBar user={user} open={open} onClose={onClose}/></Router>) ;
    expect(wrapper.find("#drawer1").at(0).props().container).toBe();
    expect(wrapper.find("#drawer1").at(0).props().open).toBeTruthy();
    expect(wrapper.find(List).length).toBe(2);
  });
});