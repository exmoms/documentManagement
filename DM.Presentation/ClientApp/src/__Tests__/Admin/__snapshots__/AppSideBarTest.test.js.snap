// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppSideBar matches snapshot 1`] = `
<Router
  history={
    Object {
      "action": "POP",
      "block": [Function],
      "createHref": [Function],
      "go": [Function],
      "goBack": [Function],
      "goForward": [Function],
      "length": 1,
      "listen": [Function],
      "location": Object {
        "hash": "",
        "pathname": "/",
        "search": "",
        "state": undefined,
      },
      "push": [Function],
      "replace": [Function],
    }
  }
>
   
  <SideBar
    onClose={[MockFunction]}
    open={true}
    user={
      Object {
        "role": "Admin",
      }
    }
  />
</Router>
`;
